confirm.c

```diff
@@ -395,50 +398,123 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
+		}
+		else if((pkt->icmpv6->icmp_type == ICMP6_TIMXCEED) &&
+				(pkt->icmpv6->icmp_code == ICMP_TIMXCEED_INTRANS)) {
+			logd(LOG_DEBUG, "%s ICMP time exceeded received\n", __func__);
+
+			struct libnet_ipv6_hdr *rip;
+			struct libnet_icmpv6_hdr *ricmp;
+			struct libnet_tcp_hdr *rtcp;
+			rip = (struct libnet_ipv6_hdr *)(pkt->payload);
+
+			if(rip->ip_nh==IPPROTO_ICMP6){
+				// Pacote interno ICMP
+				probe_type = PROBE_TYPE_ICMP;
+				ricmp = (struct libnet_icmpv6_hdr *)
+						(pkt->payload + LIBNET_IPV6_H);
+				memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
+				uint32_t flags = *(uint32_t *)(rip->ip_flags);
+				traffic_class = (flags & 0x0FF00000) >> 20;
+				flow_label = (flags & 0x000FFFFF);
+				icmpid = ntohs(ricmp->id);
+				packed_data = ntohs(ricmp->seq);
+			}
+			else if(rip->ip_nh==IPPROTO_TCP) {
+				// Pacote interno TCP
+				// TODO (rafael): draft
+				probe_type = PROBE_TYPE_TCP;
+				rtcp = (struct libnet_tcp_hdr *)(pkt->payload + LIBNET_IPV6_H);
+				memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
+				uint32_t flags = *(uint32_t *)(rip->ip_flags);
+				traffic_class = (flags & 0x0FF00000) >> 20;
+				flow_label = (flags & 0x000FFFFF);
+				packed_data = (uint16_t) (ntohl(rtcp->th_seq) & 0x0000FFFF);
+			}
+			else {
+				logd(LOG_FATAL, "%s %d: no internal TCP or ICMPv6\n", __FILE__,
+					__LINE__);
+				return NULL;
+			}
+		}
+		else if((pkt->icmpv6->icmp_type == ICMP6_DST_UNREACH) &&
+			(pkt->icmpv6->icmp_code == ICMP6_DST_UNREACH_NOPORT)){
+			// Port unreachable
+			// TODO (rafael): draft (to be tested)
+			probe_type = PROBE_TYPE_TCP;
+			struct libnet_ipv6_hdr *rip = (struct libnet_ipv6_hdr *)(pkt->payload);
+			struct libnet_tcp_hdr *rtcp = (struct libnet_tcp_hdr *)
+					(pkt->payload + LIBNET_IPV6_H);
+			memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
+			uint32_t flags = *(uint32_t *)(rip->ip_flags);
+			traffic_class = (flags & 0x0FF00000) >> 20;
+			flow_label = (flags & 0x000FFFFF);
+			packed_data = (uint16_t) (ntohl(rtcp->th_seq) & 0x0000FFFF);
+			logd(LOG_DEBUG, "%s PORT UNREACH received\n", __func__);
+		}
+		else {
+			logd(LOG_FATAL, "%s %d: unsupported ICMP type %d\n", __FILE__,
+				__LINE__, pkt->icmpv6->icmp_type);
+			return NULL;
+		}
+	}
+	else {
+		logd(LOG_FATAL, "%s %d: no TCP or ICMPv6\n", __FILE__, __LINE__);
+		return NULL;
 	}
 
 	uint8_t ttl;
 	uint8_t flowid;
 	int fixrev;
-	confirm_data_unpack(data, &ttl, &flowid, &fixrev);
+	confirm_data_unpack(packed_data, &ttl, &flowid, &fixrev);
 	assert(fixrev == 0);
 
 	struct sockaddr_storage *ptr = (struct sockaddr_storage *)&dst;
-	struct confirm_query *q = confirm_query_create6(ptr, ttl,
-			traffic_class, flow_label,
-			icmpid, flowid, NULL);
+	struct confirm_query *q;
+
+	logd(LOG_DEBUG, "received probe %d ttl %d flowid %d fixrev %d\n", probe_type, ttl, flowid, fixrev);
+
+	if(probe_type==PROBE_TYPE_ICMP){
+		q = confirm_query_create6_icmp(ptr, ttl, traffic_class, flow_label,
```

Comment:
one problem with these `if`s is that we do not fill in `traffic_class` and `flow_label` for TCP SYN-ACK and ICMP ECHOREPLY messages.

Given we only use `traffic_class` and `flow_label` to identify load balancers, can't we compute their values from the flow-ID stored in `packed_data`? The vanilla version of Paris traceroute does this. If we used the flow-ID to fill in `traffic_class` and `flow_label`, we would (1) always have this information and (2) be able to check whether the original packet was mangled in any way.

Does this make sense? Do you think we can implement it?